spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHttpAutoConfiguration.java

```diff
@@ -39,28 +48,122 @@
 @ConditionalOnBean(Tracing.class)
 @ConditionalOnProperty(name = "spring.sleuth.http.enabled", havingValue = "true", matchIfMissing = true)
 @AutoConfigureAfter(TraceWebAutoConfiguration.class)
+@EnableConfigurationProperties(SleuthHttpLegacyProperties.class)
 public class TraceHttpAutoConfiguration {
 
+	@Autowired(required = false) List<HttpClientParser> clientParsers = new ArrayList<>();
+	@Autowired(required = false) List<HttpServerParser> serverParsers = new ArrayList<>();
+	@Autowired(required = false) @ClientSampler HttpSampler clientSampler;
+	@Autowired(required = false) @ServerSampler HttpSampler serverSampler;
+
 	@Bean
 	@ConditionalOnMissingBean
 	// NOTE: stable bean name as might be used outside sleuth
 	HttpTracing httpTracing(
-			@Value("${spring.sleuth.http.legacy.enabled:false}") boolean legacyEnabled,
 			Tracing tracing,
-			TraceKeys traceKeys,
-			ErrorParser errorParser,
 			SkipPatternProvider provider
 	) {
-		if (legacyEnabled) {
-			return HttpTracing.newBuilder(tracing)
-					.clientParser(new SleuthHttpClientParser(traceKeys))
-					.serverParser(new SleuthHttpServerParser(traceKeys, errorParser))
-					.serverSampler(new SleuthHttpSampler(provider))
-					.build();
-		}
-		return HttpTracing
-				.newBuilder(tracing)
-				.serverSampler(new SleuthHttpSampler(provider))
+		return HttpTracing.newBuilder(tracing)
+				.clientParser(new CompositeClientParser(this.clientParsers))
+				.serverParser(new CompositeServerParser(this.serverParsers))
+				.clientSampler(this.clientSampler)
+				.serverSampler(new CompositeServerSampler(this.serverSampler, provider))
 				.build();
 	}
+
+	@Bean
+	@ConditionalOnProperty(name = "spring.sleuth.http.legacy.enabled", havingValue = "true")
+	HttpClientParser sleuthHttpClientParser(TraceKeys traceKeys) {
+		return new SleuthHttpClientParser(traceKeys);
+	}
+
+	@Bean
+	@ConditionalOnProperty(name = "spring.sleuth.http.legacy.enabled", havingValue = "true")
+	HttpServerParser sleuthHttpServerParser(TraceKeys traceKeys, ErrorParser errorParser) {
+		return new SleuthHttpServerParser(traceKeys, errorParser);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(name = "sleuthClientSampler")
+	HttpSampler sleuthClientSampler() {
+		return HttpSampler.TRACE_ID;
+	}
+}
+
+class CompositeClientParser extends HttpClientParser {
```

Comment:
composite is trickier than just allowing someone to override the type. For example, they can't remove tags this way. I might start with just letting people override it. If composition becomes common we can add